project5/3700kvstore

```diff
@@ -275,24 +338,32 @@ class Server:
         prev_log_idx = msg['prev_log_idx']
         prev_log_term = msg['prev_log_term']
 
+        def reject(string):
+            self.info('Failing {}.append_entries ({})'.format(msg['src'], string))
+            self.send_fail(msg)
+
         # If 'Leader' has a term equal to or greater than term, they
         # had to be properly elected and must be safe to follow
         if msg['term'] < self.current_term:
-            logging.info('Recieved out-of-term message')
-            self.send_fail(msg)
+            reject('term {}<{}'.format(self.current_term, msg['term']))
         else:
             self.transition_to_state(State.FOLLOWER, msg)
             self.reset_timeouts()
 
             if prev_log_idx > self.last_log_idx:
-                logging.info('Prev log index greater than max log index')
-                self.send_fail(msg)
+                reject('index {}>{}'.format(prev_log_idx, self.last_log_idx))
             elif self.log[prev_log_idx]['term'] != prev_log_term:
-                logging.info('prev term doesn\'t match {} {}'.format(self.log[prev_log_idx]['term'], prev_log_term))
-                self.send_fail(msg)
+                reject('prev term {}!={}'.format(self.log[prev_log_idx]['term'],
+                        prev_log_term))
             else:
-                self.log = self.log[:prev_log_idx + 1]
-                self.log.extend(msg['entries'])
+                if msg['entries']:
+                    last = prev_log_idx + 1
+                    if last <= self.last_log_idx:
+                        self.info('Deleting entries [{}:]'.format(last))
+                        del self.log[last:]
+                    self.info('Appending {} entries ({})'.format(len(msg['entries']),
+                            [e['MID'] for e in msg['entries']]))
+                    self.log.extend(msg['entries'])
                 self.commit_idx = min(self.last_log_idx, msg['leader_commit'])
```

Comment:
The Raft paper says that we should only set the commit index to the min if `if msg['leader_commit'] > self.commit_idx`.

Reply:
that makes sense, guess we could run into a problem where  this caused  us to walk back what we said was committed.

I can't help but think we shouldn't run into a situation like this since RAFT guarantees in elections the leader has the longest log and thus leader commit should never decrease